### PR TITLE
Migration of metricset.* fields to ECS

### PR DESCRIPTION
**Pull request by ruflin**
* Add `event.dataset` as `{module}.{metricset}` but keep `metricset.name` around for backward compatiblity.
* Add `event.duration` which comes from `metricset.rtt` but is now in nano instead of micro seconds. `event.duration` will be introduce in 6.x already.
* Rename `metricset.module` to `event.module`
* Rename `metricset.host` to `service.address`
* Removed `metricset.namespace` as this becomes `event.dataset` and should now always be set.

Further changes

* Updated some of the data.json files as example
* Migration file updated
* Old field definitions removed
* System tests updated

### CODE REVIEW

File: metricbeat/mb/event.go [hunk 1]
```diff
@@ -109,27 +109,26 @@ func (e *Event) BeatEvent(module, metricSet string, modifiers ...EventModifier)
...
-		"module": module,
+	e := common.MapStr{
+		"event": common.MapStr{
+			"dataset": metricset,
+			"module":  module,
+		},
 	}
 	if event.Host != "" {
-		info["host"] = event.Host
+		e.Put("service.hostname", event.Host)
```
Comment (ruflin) [inline metricbeat/mb/event.go]: @webmat To make things a bit more interesting, `Host` here contains also the Port number. For example `elasticsearch:9200`.

File: metricbeat/mb/event.go [hunk 2]
```diff
@@ -109,27 +109,26 @@ func (e *Event) BeatEvent(module, metricSet string, modifiers ...EventModifier)
...
 	}
 	if event.Took > 0 {
-		info["rtt"] = event.Took / time.Microsecond
+		e.Put("event.duration", event.Took/time.Microsecond)
```
Comment (andrewkroh) [inline metricbeat/mb/event.go]: This is described as being in nanoseconds in ECS.

Comment (ruflin) [reply metricbeat/mb/event.go]: @spinscale What will happen if we send nano seconds data to ES versions prior 7?

Comment (ruflin) [reply metricbeat/mb/event.go]: Good point, will convert it.

Also not related to nano second support in ES as it's not a timestamp.

Comment (ruflin) [reply metricbeat/mb/event.go]: @andrewkroh As this is a breaking change, I wonder if we want to keep the old field around?

Other option would be to introduce the new field in nano seconds already in 6.6, deprecate the other one and then remove it in 7.0?

File: dev-tools/ecs-migration.yml [hunk 3]
```diff
@@ -90,3 +90,26 @@
...
+  alias: true
+  copy_to: false
+
+# TODO: Is it really an alias? Or will we change the unit?
+- from: metricset.rrt
+  to: event.duration
+  alias: true
+  copy_to: false
+
+# TODO: This becomes port and domain in servce?
```
Comment (ruflin) [inline dev-tools/ecs-migration.yml]: @webmat Feedback welcome on this one.

Comment (webmat) [reply dev-tools/ecs-migration.yml]: Hmmm, service doesn't have domain, ip or port.

Is the intent to represent the address exposed by a service? E.g.

```YAML
service.name: sshd
service.port: 22
service.ip: 10.10.10.10
service.domain: (hostname on the local network)
```

File: dev-tools/ecs-migration.yml [hunk 4]
```diff
@@ -90,3 +90,26 @@
...
+  copy_to: false
+
+- from: metricset.module
+  to: event.module
+  alias: true
+  copy_to: false
+
+# TODO: Is it really an alias? Or will we change the unit?
+- from: metricset.rrt
+  to: event.duration
```
Comment (webmat) [inline dev-tools/ecs-migration.yml]: Yeah if this one is changing by a factor of 1 000, I think we can't make this an alias. We'll have to manage the deprecation a bit more manually. Example:

6.x

- Mark `metricset.rrt` as deprecated
- Keep populating `metricset.rrt` field with microseconds
- Start populating `event.duration` field with nanoseconds
- Convert all visualizations to use new field
  - Make sure to correctly set Kibana Index Pattern units for new field

7.x

- remove `metricset.rtt`

File: CHANGELOG.asciidoc [hunk 5]
```diff
@@ -22,6 +22,7 @@ https://github.com/elastic/beats/compare/v7.0.0-alpha1...master[Check the HEAD d
 *Journalbeat*
 
 *Metricbeat*
+- event.duration is now in nano and not micro seconds anymore.
```
Comment (andrewkroh) [inline CHANGELOG.asciidoc]: `s/micro seconds/microseconds/`

File: metricbeat/mb/event.go (hunk 2 above)
Comment (andrewkroh) [reply metricbeat/mb/event.go]: The plan you describe sounds good to me. Have both in 6.6 and remove rtt in 7.0.

File: dev-tools/ecs-migration.yml (hunk 3 above)
Comment (ruflin) [reply dev-tools/ecs-migration.yml]: It is the host/hostname that the user configured to connect to the service to fetch the information.

**Comment by ruflin** [comment]
Here is the 6.x PR to introduce `event.duration` and `event.dataset`: https://github.com/elastic/beats/pull/9393

File: dev-tools/ecs-migration.yml (hunk 3 above)
Comment (webmat) [reply dev-tools/ecs-migration.yml]: Ok, removed from here, got it.

I was thinking of introducing a field for these ambiguous addresses as `address`. It's an extremely generic word, it doesn't clash with any of the words used in the RFC that have a *somewhat* specific meaning like "host", "hostname", or "authority". `address` also kinda works for situations where e.g. NGINX or HAProxy shove a unix socket in the "IP" field.

I think "address" could also work for DSNs like the mysql example above.

File: dev-tools/ecs-migration.yml [hunk 6]
```diff
@@ -367,3 +366,29 @@
...
+  alias: true
+
+- from: metricset.rrt
+  to: event.duration
+  alias: false
+  comment: Unit changed, old field will stay around if needed
+  breaking: true
+
+- from: metricset.host
+  to: service.uri
```
Comment (ruflin) [inline dev-tools/ecs-migration.yml]: @ruflin Note for myself: This should be `service.address`

**Comment by ruflin** [comment]
jenkins, test this

**Comment by webmat** [comment]
jenkins, test this

File: metricbeat/mb/event.go [hunk 7]
```diff
@@ -116,27 +117,32 @@ func (e *Event) BeatEvent(module, metricSet string, modifiers ...EventModifier)
...
+			"module":  module,
+		},
+		// TODO: This should only be sent if migration layer is enabled
+		"metricset": common.MapStr{
+			"name": metricset,
+		},
 	}
 	if event.Host != "" {
-		info["host"] = event.Host
+		e.Put("service.address", event.Host)
```
Comment (webmat) [inline metricbeat/mb/event.go]: Isn't `service.address` meant to be a remote address?

File: metricbeat/mb/module/example_test.go [hunk 8]
```diff
@@ -64,7 +64,7 @@ func ExampleWrapper() {
 	go func() {
 		defer wg.Done()
 		for event := range output {
-			event.Fields.Put("metricset.rtt", 111)
+			event.Fields.Put("event.duration", 111)
```
Comment (webmat) [inline metricbeat/mb/module/example_test.go]: Oof! 111 ns is a short-lived event 😆

File: metricbeat/module/system/diskio/_meta/data.json [hunk 9]
```diff
@@ -1,60 +1,35 @@
...
-                "write": {
-                    "per_sec": {
-                        "bytes": 0
-                    },
-                    "request": {
-                        "merges_per_sec": 0,
-                        "per_sec": 0
-                    },
-                    "await": 0
-                }
```
Comment (webmat) [inline metricbeat/module/system/diskio/_meta/data.json]: Is the removal of this big chunk expected?

File: metricbeat/module/system/process/_meta/data.json [hunk 10]
```diff
@@ -1,198 +1,46 @@
...
-                            "bytes": 0
-                        },
-                        "unevictable": {
-                            "bytes": 0
-                        }
-                    }
-                },
-                "path": "/docker/d88e67bb6961a5bb70c1c1c48094c6030e43768eed91e827f437111888f9967e"
-            },
-            "cmdline": "go test -tags=integration github.com/elastic/beats/metricbeat/module/... -data",
```
Comment (webmat) [inline metricbeat/module/system/process/_meta/data.json]: Is the removal of this big chunk expected?

File: metricbeat/module/system/process/_meta/data.json [hunk 11]
```diff
@@ -1,198 +1,46 @@
...
                 }
             },
-            "cwd": "/go/src/github.com/elastic/beats/metricbeat",
-            "fd": {
-                "limit": {
-                    "hard": 1048576,
-                    "soft": 1048576
-                },
-                "open": 8
-            },
```
Comment (webmat) [inline metricbeat/module/system/process/_meta/data.json]: Is the removal of this comparatively small chunk expected?

File: metricbeat/mb/event.go (hunk 7 above)
Comment (ruflin) [reply metricbeat/mb/event.go]: It can be both. It is what the user configured.

File: metricbeat/mb/module/example_test.go (hunk 8 above)
Comment (ruflin) [reply metricbeat/mb/module/example_test.go]: It's a hardcoded value for testing.

File: metricbeat/module/system/diskio/_meta/data.json (hunk 9 above)
Comment (ruflin) [reply metricbeat/module/system/diskio/_meta/data.json]: There are automatically generated files. No need to worry about these. These are only examples for the docs.

File: metricbeat/module/system/process/_meta/data.json (hunk 10 above)
Comment (ruflin) [reply metricbeat/module/system/process/_meta/data.json]: Same here, ignore all `data.json` changes.

**Pull request merged**